resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java

```diff
@@ -620,6 +629,34 @@ public String toString(){
         };
     }
 
+    /**
+     * Returns a supplier of type Future which is decorated by a CircuitBreaker.
+     *
+     * @param circuitBreaker the CircuitBreaker
+     * @param supplier       the original supplier
+     * @param <T>            the type of the returned Future's result
+     * @return a supplier which is decorated by a CircuitBreaker.
+     */
+    static <T> Supplier<Future<T>> decorateFuture(CircuitBreaker circuitBreaker, Supplier<Future<T>> supplier) {
+       return () -> {
+           if (!circuitBreaker.tryAcquirePermission()) {
```

Comment:
I think when I invoke a method which returns a Future (and which does not throw any checked exceptions), I wouldn't expect to get a RuntimeException like `CallNotPermittedException`. I would expect to get the exception inside of an `ExecutionException` when I call `future.get`.  We should direclty return a Future completed exceptionallym when the acquire permission was rejected.

```
final CompletableFuture<T> promise = new CompletableFuture<>();
promise.completeExceptionally(e);
return promise;
```

Reply:
Following are thoughts behind the solution. I am flexible here.. 

1. There are certain gotchas with using `CompletableFuture` with `Future` type.  Wrapping exception inside `promise.completeExceptionally(e);` will wrap everything inside `ExecutionException` and it will break `Future` contract for example

```
final CompletableFuture<T> promise = new CompletableFuture<>();
try{
  T v = future.get(5, SECONDS);
  promise.complete(v);
} catch(ExecutionException e){
  promise.completeExceptionally(e.getCause()); //Works
} catch (CancellationException e){
  promise.cancel(true); //Works
} catch (TimeoutException | InterruptedException e){
  promise.completeExceptionally(e); //Breaks Contract as TimeoutException & InterruptedException wraps inside ExecutionException
}
return promise;
```

Also, the downside of this approach is that we have to invoke `get()` to find results where as with `CircuitBreakerFuture` we just decorate future and let caller invoke it.

2. Regarding `CallNotPermittedException` exception there are two scenario here, Supplier fails before `Future` is even returned and Second, invoking `Future` fails. 

::Example: Supplier fails before `Future` is even returned
```
Future<String> work = () -> {
    Supplier<String> t = svc.execute() //this blows up!
    return executorService.submit(t);
}
Future<String> decorateFuture = circuitBreaker.decorateFuture(work);
```
The above scenario would not be count towards CB if we do not handle it outside `Future`

::Example: Invoking `Future` fails: this would handle in `CircuitBreakerFuture::get` method.

Reply:
What about this
```
    static <T> Supplier<Future<T>> decorateFuture(CircuitBreaker circuitBreaker, Supplier<Future<T>> supplier) {
        return () -> {
            if (!circuitBreaker.tryAcquirePermission()) {
                CompletableFuture<T> promise = new CompletableFuture<>();
                promise.completeExceptionally(CallNotPermittedException.createCallNotPermittedException(circuitBreaker));
                return promise;
            } else {
                final long start = System.nanoTime();
                try {
                    return new CircuitBreakerFuture<>(circuitBreaker, supplier.get(), start);
                } catch (Exception e) {
                    long durationInNanos = System.nanoTime() - start;
                    circuitBreaker.onError(durationInNanos, TimeUnit.NANOSECONDS, e);
                    throw e;
                }
            }
        };
    }

    final class CircuitBreakerFuture<T> implements Future<T> {
        private CircuitBreaker circuitBreaker;
        private Future<T> future;
        private long start;
        private AtomicBoolean processed = new AtomicBoolean(false);

        CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future, long start) {
            Objects.requireNonNull(future, "Non null Future is required to decorate");
            this.circuitBreaker = circuitBreaker;
            this.future = future;
            this.start = start;
        }

        @Override
        public boolean cancel(boolean mayInterruptIfRunning) {
            return future.cancel(mayInterruptIfRunning);
        }

        @Override
        public boolean isCancelled() {
            return future.isCancelled();
        }

        @Override
        public boolean isDone() {
            return future.isDone();
        }

        @Override
        public T get() throws InterruptedException, ExecutionException {
            try {
                T v = future.get();
                if(processed.compareAndSet(false, true)){
                    long durationInNanos = System.nanoTime() - start;
                    circuitBreaker.onSuccess(durationInNanos, TimeUnit.NANOSECONDS);
                }
                return v;
            } catch (CancellationException | InterruptedException e) {
                circuitBreaker.releasePermission();
                throw e;
            } catch (Exception e) {
                if(processed.compareAndSet(false, true)){
                    long durationInNanos = System.nanoTime() - start;
                    circuitBreaker.onError(durationInNanos, TimeUnit.NANOSECONDS, e.getCause());
                }
                throw e;
            }
        }

        @Override
        public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
            try {
                T v = future.get(timeout, unit);
                long durationInNanos = System.nanoTime() - start;
                circuitBreaker.onSuccess(durationInNanos, TimeUnit.NANOSECONDS);
                return v;
            } catch (CancellationException | InterruptedException e) {
                circuitBreaker.releasePermission();
                throw e;
            } catch (Exception e) {
                if(processed.compareAndSet(false, true)){
                    long durationInNanos = System.nanoTime() - start;
                    circuitBreaker.onError(durationInNanos, TimeUnit.NANOSECONDS, e.getCause());
                }
                throw e;
            }
        }
    }
```

Reply:
Given that we're OK with stopping the clock at the first `get` if necessary, I like this solution. Would suggest factoring the "only call to the CircuitBreaker once" and "pass in execution time" into their own utility classes, e.g.:

```java
final class OnceSupplier<T> {
    private final T val;
    private final AtomicBoolean hasRun = new AtomicBoolean(false);
    public OnceSupplier(T val) { this.val = val; }
    public void supplyOnce(Consumer<T> consumer) {
        if (hasRun.compareAndSet(false, true)) {
            consumer.accept(val);
        }
    }
}

final class NanosecondTimer {
    private final long startNanos;
    public NanosecondTimer(long startNanos) { this.startNanos = startNanos; }
    public void runWithTime(LongConsumer consumer) {
        consumer.accept(System.nanoTime() - startNanos);
    }
}

final class CircuitBreakerFuture<T> implements Future<T> {
    private Future<T> future;
    private OnceSupplier<CircuitBreaker> onceToCircuitBreaker;
    private NanosecondTimer durationTimer;

    public CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future, long start) {
        this.future = future;
        this.onceToCircuitBreaker = new OnceSupplier<>(circuitBreaker);
        this.durationTimer = new NanosecondTimer(start);
    }

    ...

    @Override
    public T get() throws InterruptedException, ExecutionException {
        try {
            T v = future.get();
            onceToCircuitBreaker.supplyOnce(circuitBreaker ->
                durationTimer.runWithTime(durationInNanos ->
                    circuitBreaker.onSuccess(durationInNanos, TimeUnit.NANOSECONDS)));
        } catch (CancellationException | InterruptedException e) {
            onceToCircuitBreaker.supplyOnce(CircuitBreaker::releasePermission);
            throw e;
        } catch (Exception e) {
            onceToCircuitBreaker.supplyOnce(circuitBreaker ->
                durationTimer.runWithTime(durationInNanos ->
                    circuitBreaker.onError(durationInNanos, TimeUnit.NANOSECONDS, e)));
            throw e;
        }
    }

    ...
}
```